### PR TITLE
Make Serializer::Mutable very mutable

### PR DESCRIPTION
**Pull request by dboehmer**
As part of my working on fixing GH #795 I see an opportunity to make this module capable of using mixed input/output formats.

This might not be the most common case but while rewriting the test it just happened and here you have the complete feature incl. test and updated documentation. Hope you like it.

### CODE REVIEW

**Comment by shumphrey** [comment]
The thing is, Serializers (Mutable in particular) cannot deserialize without the request object (or specifically the content_type).
Serializers do not have request at the point of deserializing the request body.
I am not keen on Serializer holding on to a reference to request. I'd much rather just pass request or content_type as a parameter to deserialize, although using the existing $options which deserialize currently takes breaks other serializers

**Comment by xsawyerx** [comment]
Where are we with this?

**Comment by dboehmer** [comment]
As I wrote before I'd change my PR to match some spec how it should work but this issue is still missing a proposed spec.

File: t/serializer_mutable.t [hunk 1]
```diff
@@ -1,90 +1,162 @@
...
 
-my $app = MyApp->to_app;
-ok( is_coderef($app), 'Got app' );
+isa_ok my $app = MyApp->to_app => 'CODE', "app";
+
+use Plack::Middleware::TrafficLog;
+$app = Plack::Middleware::TrafficLog->wrap(
+    $app,
+    logger => sub { return; warn "@_" },
+    with_body => 1,    # TODO find out why this fixes the test
```
Comment (dboehmer) [inline t/serializer_mutable.t]: Has anyone an idea why this fixes the test? Without that line I get the following error and I fail to debug where it originates.

```
t/serializer_mutable.t .. 1/10 
    #   Failed test 'response status code'
    #   at t/serializer_mutable.t line 141.
    #          got: '500'
    #     expected: '200'

    #   Failed test 'response content type'
    #   at t/serializer_mutable.t line 143.
    #          got: 'text/plain'
    #     expected: 'text/x-yaml'

    #   Failed test 'response content'
    #   at t/serializer_mutable.t line 158.
    #          got: 'Internal Server Error
    # 
    # Bad Content-Length: maybe client disconnect? (6 bytes remaining) at /home/USER/perl5/perlbrew/perls/perl-5.22.1/lib/site_perl/5.22.1/Plack/Request.pm line 260.
    # '
    #     expected: '---
    # foo: bar
    # '
    # Looks like you failed 3 tests of 3.

#   Failed test 'unsupported 'Content-Type' and supported 'Accept' header'
#   at t/serializer_mutable.t line 160.
# Looks like you failed 1 test of 10.
t/serializer_mutable.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/10 subtests 

Test Summary Report
-------------------
t/serializer_mutable.t (Wstat: 256 Tests: 10 Failed: 1)
  Failed test:  8
  Non-zero exit status: 1
Files=1, Tests=10,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.30 cusr  0.05 csys =  0.37 CPU)
Result: FAIL
```

**Comment by dboehmer** [comment]
I pushed a complete rewrite of `Serializer::Mutable`. POD still needs to be updated. Please take a look into the test and module code if the approach is acceptable for integration once completed. See code comment for unresolved issue (currently working due to weird workaround).

**Comment by dboehmer** [comment]
@xsawyerx @veryrusty @shumphrey Sorry to bother you but I feel like nobody got a notification for my update from September 29. Can you please review or write what's missing or when someone will have time to do that?

**Comment by veryrusty** [comment]
I think #1457 covers everything this PR was to achieve.

( Sorry its taken years to get that far :( )

**Pull request closed without merging**